### PR TITLE
Yaml optimization : new scene hierarchy cache 

### PR DESCRIPTION
**Pull request by krasnotsvetov**

### CODE REVIEW

File: resharper/resharper-unity/src/Rider/UnityEditorFindUsageResultCreator.cs [hunk 1]
```diff
@@ -153,7 +158,12 @@ public IUnityYamlReference Build(FindResult result)
 
             public FindExecution Merge(IUnityYamlReference data)
             {
-                var request = CreateRequest(mySolutionDirectoryPath, mySceneProcessor, data.ComponentDocument, data == mySelectedReference);
+                var sourceFile = data.ComponentDocument.GetSourceFile();
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Rider/UnityEditorFindUsageResultCreator.cs]: we need to check if the sourceFile is valid

File: resharper/resharper-unity/src/Rider/UnityYamlOccurenceProvider.cs [hunk 2]
```diff
@@ -25,6 +26,7 @@ public IOccurrence MakeOccurrence(FindResult findResult)
                 
                 if (reference != null)
                 {
+                    var cache = reference.ComponentDocument.GetSolution().GetComponent<UnitySceneDataCache>();
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Rider/UnityYamlOccurenceProvider.cs]: todo for 19.3. [OccurenceProviders] should be composed via SolutionFeaturePartsContainer, so such components could just inject solution and do not do .GetComponent on every occurrence create.

File: resharper/resharper-unity/src/Yaml/Daemon/UsageChecking/YamlCollectUsagesPsiFileProcessor.cs [hunk 3]
```diff
@@ -22,24 +24,39 @@ public bool InteriorShouldBeProcessed(ITreeNode element, UsageData context)
...
 
-                context.SetElementState(declaredElement, UsageState.USED_MASK);
-                context.AddUsage(declaredElement, ReferenceCounter.Id);
-                context.AddUsage(declaredElement, UnityEditorUsageCounter.Id);
+                usageData.SetElementState(declaredElement, UsageState.USED_MASK);
+                usageData.AddUsage(declaredElement, ReferenceCounter.Id);
+                usageData.AddUsage(declaredElement, UnityEditorUsageCounter.Id);
+            }
+            
+            foreach (var (_, swaElement) in context.UsageData.SwaExtensionsData)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Daemon/UsageChecking/YamlCollectUsagesPsiFileProcessor.cs]: do we really need to iterate through SwaExtensionsData *two times*??

File: resharper/resharper-unity/src/Yaml/Psi/Caches/MetaFileGuidCache.cs [hunk 4]
```diff
@@ -67,6 +67,16 @@ public override object Build(IPsiSourceFile sourceFile, bool isStartup)
...
             var document = yamlFile.Documents.FirstOrDefault();
+
+            var guid = GetGuid(document);
+            if (guid != null)
+                return new MetaFileCacheItem(guid);
+            
+            return null;
+        }
+
+        public static string GetGuid(IYamlDocument document)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/MetaFileGuidCache.cs]: yet another public static method without any threading/lock requirements

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                name = null;
+            
+            myParts.Push(name ?? "...");
+            
+            int? rootOrder = null;
+            if (modifications != null)
+            {
+                rootOrder = modifications.GetRootIndex(transformHierarchyElement.Id.WithGuid(null));
+            }
+            if (rootOrder == null)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs]: else if?

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        void ConsumeGameObject(GameObjectHierarchyElement gameObject, TransformHierarchyElement transformHierarchyElement, ModificationHierarchyElement modifications);
+    }
+    
+    public class UnityPathCachedSceneConsumer : IUnityCachedSceneProcessorConsumer
+    {
+        public List<string> NameParts => myParts.ToList();
+        public List<int> RootIndexes => myIndex.ToList();
+
+        private readonly Stack<string> myParts = new Stack<string>();
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs]: А зачем нам тут стэк? По коду мы пушим туда только один раз, в кторе, и всё.

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/PropertyValueLocalCache.cs [hunk 7]
```diff
@@ -140,6 +152,8 @@ public override int GetHashCode()
             }
         }
         
+        
+        // Different indexes for queries from local stage. All queries should take O(1)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/PropertyValueLocalCache.cs]: 'indices' not 'indexes'

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/SceneHierarchy.cs [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+using JetBrains.ReSharper.Psi.Parsing;
+using JetBrains.ReSharper.Psi.Web.WebConfig;
+using JetBrains.Serialization;
+using JetBrains.Text;
+using JetBrains.Util;
+using JetBrains.Util.Collections;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Yaml.Psi.Caches.UnityEditorPropertyValues
+{
+    public class SceneHierarchy
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/SceneHierarchy.cs]: almost everything is public here, incl public static APIs. not requirements for the threading. 
If this is a free threaded code - these dictionaries won't work. If it is a UI thread only code - mark it please with assertion.

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/SceneHierarchy.cs [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+                   referenceValues.GetValueSafe(UnityYamlConstants.CorrespondingSourceObjectProperty2017);
+        }
+        
+        private FileID GetPrefabInstanceId(Dictionary<string, FileID> referenceValues)
+        {
+            return referenceValues.GetValueSafe(UnityYamlConstants.PrefabInstanceProperty) ??
+                   referenceValues.GetValueSafe(UnityYamlConstants.PrefabInstanceProperty2017);
+        }
+
+        public void AddPrefabModification(IBuffer buffer)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/SceneHierarchy.cs]: I will pray for this algorithm, cause I don't really understand what's going on here. 🕯

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneData.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+using JetBrains.Util;
+using JetBrains.Util.PersistentMap;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Yaml.Psi.Caches.UnityEditorPropertyValues
+{
+    public class UnitySceneData
+    {
+        private static readonly StringSearcher ourPrefabModificationSearcher = new StringSearcher("!u!1001", true);
+        
+        public readonly OneToListMap<MonoBehaviourProperty, MonoBehaviourPropertyValue> PropertiesData;
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneData.cs]: multithreading?

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneData.cs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                (
+                    new UniversalMarshaller<MonoBehaviourProperty>(MonoBehaviourProperty.ReadFrom,
+                        MonoBehaviourProperty.WriteTo),
+                    new UniversalMarshaller<MonoBehaviourPropertyValue>(
+                        MonoBehaviourPropertyValueMarshaller.Read,
+                        MonoBehaviourPropertyValueMarshaller.Write),
+                    n => new OneToListMap<MonoBehaviourProperty, MonoBehaviourPropertyValue>(n)
+                );
+
+        public static UnitySceneData Build(IUnityYamlFile file)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneData.cs]: readlock? threading?

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataCache.cs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        public readonly UnitySceneDataLocalCache UnitySceneDataLocalCache;
+        
+        public UnitySceneDataCache(Lifetime lifetime, MetaFileGuidCache metaFileGuidCache, IPersistentIndexManager persistentIndexManager)
+            : base(lifetime, persistentIndexManager, UnitySceneData.Marshaller)
+        {
+            UnitySceneDataLocalCache = new UnitySceneDataLocalCache(metaFileGuidCache);
+        }
+        
+        protected override bool IsApplicable(IPsiSourceFile sourceFile)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataCache.cs]: check that sourceFile is valid

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        {
+            var sourceFile = startComponent.GetSourceFile();
+            var anchor = UnitySceneDataUtil.GetAnchorFromBuffer(startComponent.GetTextAsBuffer());
+            if (sourceFile == null || anchor == null)
+                return;
+            
+            ProcessSceneHierarchyFromComponentToRoot(sourceFile, new FileID(null, anchor),  consumer);
+        }
+        
+        public void ProcessSceneHierarchyFromComponentToRoot(IPsiSourceFile sourceFile, string anchor, IUnityCachedSceneProcessorConsumer consumer)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs]: either assert or check that the sourceFile is valid

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+                    var unityModule = currentFile.PsiModule as UnityExternalFilesPsiModule;
+                    if (unityModule == null)
+                        return;
+                    
+                    if (!unityModule.NotNull("externalFilesModuleFactory.PsiModule != null")
+                        .TryGetFileByPath(prefabSourceFilePaths.First(), out var prefabSourceFile))
+                        return;
+                        
+
+                    var prefabStartGameObject = mySceneElements.GetValues(new SceneElementId(prefabSourceFile, correspondingId.WithGuid(null))).FirstOrDefault();
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs]: why FirstOrDefault here?

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+                    
+                    if (prefabStartGameObject == null)
+                        return; // TODO [vkrasnotsvetov] 19.3 Handle case, when prefab contains prefab which contains prefab
+                    
+                    if (!prefabStartGameObject.IsStripped)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs]: can this piece of code be simplified? I.t. ifs can be probably merged and apparently the whole code arrangement could be done more nicely

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataUtil.cs [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+using JetBrains.ReSharper.Psi.JavaScript.Util.Literals;
+using JetBrains.ReSharper.Psi.Parsing;
+using JetBrains.Text;
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Yaml.Psi.Caches.UnityEditorPropertyValues
+{
+    public static class UnitySceneDataUtil
+    {
+        public static void ExtractSimpleAndReferenceValues(IBuffer buffer, Dictionary<string, string> simpleValues, Dictionary<string, FileID> referenceValues)
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataUtil.cs]: again. Public api. threading. lock.

File: resharper/resharper-unity/src/Yaml/Rider/Host/Feature/UnityYamlExtraGroupingRulesProvider.cs [hunk 17]
```diff
@@ -95,9 +96,13 @@ public override RdUsageGroup CreateModel(IOccurrence occurrence, IOccurrenceBrow
                 if (occurrence is ReferenceOccurrence referenceOccurrence &&
                     referenceOccurrence.PrimaryReference is IUnityYamlReference reference)
                 {
-                    return CreateModel(
-                        UnityObjectPsiUtil.GetGameObjectPathFromComponent(mySceneProcessor,
-                            reference.ComponentDocument));
+                    var document = reference.ComponentDocument;
```
Comment (kskrygan) [inline resharper/resharper-unity/src/Yaml/Rider/Host/Feature/UnityYamlExtraGroupingRulesProvider.cs]: sourceFile.IsValid()

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs (hunk 5 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs]: `modifications.GetRootIndex` can return ` null` value.

File: resharper/resharper-unity/src/Rider/CodeInsights/UnityCodeInsightFieldUsageProvider.cs [hunk 18]
```diff
@@ -157,20 +157,8 @@ public override void OnClick(CodeInsightsHighlighting highlighting, ISolution so
...
-
-                                    var document = file.FindDocumentByAnchor(value.Value.MonoBehaviour);
-                                    if (document == null)
-                                        return;
-                                    var request = UnityEditorFindUsageResultCreator.CreateRequest(
-                                        solution.SolutionDirectory, mySceneProcessor, document);
-                                    myUnityHost.PerformModelAction(t => t.ShowGameObjectOnScene.Fire(request));
-                                    UnityFocusUtil.FocusUnity(myUnityHost.GetValue(t => t.UnityProcessId.Value));
-                                });
+                                UnityEditorFindUsageResultCreator.CreateRequestAndShow(myUnityHost, solution.SolutionDirectory, myUnitySceneDataLocalCache,
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Rider/CodeInsights/UnityCodeInsightFieldUsageProvider.cs]: This operation is not heavy (visiting dictionary several times).I have removed focusing window from readlock. Als

File: resharper/resharper-unity/src/Rider/CodeInsights/UnityCodeInsightFieldUsageProvider.cs [hunk 19]
```diff
@@ -238,7 +226,7 @@ private bool ShouldShowUnknownPresentation(MonoBehaviourPropertyValue value, Uni
             var guid = result.Value.guid;
             var propertyName = result.Value.propertyName;
 
-            var cache = solution.GetComponent<UnityPropertyValueCache>();
+            var cache = solution.GetComponent<UnitySceneDataCache>().UnitySceneDataLocalCache;
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Rider/CodeInsights/UnityCodeInsightFieldUsageProvider.cs]: This code is invoked from element problem analyzers

File: resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs (hunk 6 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/IUnityCachedSceneProcessorConsumer.cs]: hm? It is not true. Each time we visit hierarchy element we push to stacks. We are visiting them in inverse order

**Pull request merged**